File: frontend/src/types/prediction-api.ts
/** Mirrors backend/app/schemas/prediction_api.py — keep in sync. */
import type { Hotspot } from "@/types/prediction";

export interface PredictionOut {
  id: string;
  image_id: string;
  status: "completed" | "failed";
  schema_version: string;
  co2_emission_tonnes_per_year: number | null;
  confidence: number | null; // v1 only - see data_source for v2 predictions
  hotspots: Hotspot[] | null;
  inference_time_ms: number | null;
  created_at: string;
  // v2 fields - null on v1 rows. When data_source is set, don't assume
  // co2_emission_tonnes_per_year is present - it's null for "cnn_proxy"/
  // "unavailable" (no direct CO2 measurement for that location).
  data_source: "oco3_estimated" | "cnn_proxy" | "unavailable" | null;
  detection_confidence: number | null;
  co2_ppm_enhancement: number | null;
  co2_estimate_low: number | null;
  co2_estimate_high: number | null;
  // True only when co2_emission_tonnes_per_year came from the research
  // repo's CEA-ground-truth-corrected estimate, not this platform's own
  // placeholder mass-balance formula. Null on v1 rows and on v2 rows
  // where data_source isn't "oco3_estimated".
  ground_truth_validated: boolean | null;
  model_version: string | null;
  image_filename: string | null;
}

export interface PredictionListOut {
  items: PredictionOut[];
  total: number;
  page: number;
  page_size: number;
}

File: backend/app/schemas/prediction_api.py
"""Public API schemas for predictions (distinct from the internal ML contract
in schemas/prediction.py, which is the backend <-> ML service boundary)."""
import uuid
from datetime import datetime

from pydantic import BaseModel

from app.schemas.prediction import Hotspot


class RunPredictionRequest(BaseModel):
    image_id: uuid.UUID


class PredictionOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    image_id: uuid.UUID
    status: str
    schema_version: str
    co2_emission_tonnes_per_year: float | None
    confidence: float | None  # v1 only - see data_source for v2 predictions
    hotspots: list[Hotspot] | None
    inference_time_ms: float | None
    created_at: datetime
    # v2 fields - null on v1 rows. When data_source is present, the
    # frontend should branch on it rather than assuming
    # co2_emission_tonnes_per_year is always populated (it's null for
    # "cnn_proxy"/"unavailable" - see PredictionResultV2's docstring).
    data_source: str | None = None
    detection_confidence: float | None = None
    co2_ppm_enhancement: float | None = None
    co2_estimate_low: float | None = None
    co2_estimate_high: float | None = None
    # populated by service, not ORM - heatmap_url comes from the heatmap_key
    # column, which despite its name stores whatever URL the ML service
    # returned (not a backend-owned storage key).
    model_version: str | None = None
    image_filename: str | None = None
    heatmap_url: str | None = None


class PredictionListOut(BaseModel):
    items: list[PredictionOut]
    total: int
    page: int
    page_size: int
